### ISSUE TITLE
turning colors on/off

### ISSUE THREAD

**Issue by marcotrosi** [charmbracelet/lipgloss#445]
Hello everyone,

this is maybe more a question, that becomes a Feature Request if there is no simple solution yet, but at least I couldn't find anything in the documentation.

As you know some command line tools have something like a --no-color option to turn off all escape sequences for colors etc., or some do it automatically when they detect the tools is getting used in a pipe chain.

My first question would be "how do you implement such feature and is there some switch or other easy solution available in lipgloss?"

**Describe alternatives you've considered**
Currently I'm overwriting the Styles like this (just the basic concept, the code looks actually a bit different) ...

    StyleBigger  = lipgloss.NewStyle().Foreground(lipgloss.Color("10"))
    StyleSmaller = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
    
    if NoColor {
        StyleBigger  = lipgloss.NewStyle()
        StyleSmaller = lipgloss.NewStyle()
        ...
    }

    Result = StyleBigger.Render(myString)

I want to avoid wrapping everything and having if-else statements everywhere.

**Describe the solution you'd like**
I thought maybe something simple like the following concept would be simple ...

    StyleBigger  = lipgloss.NewStyle().Foreground(lipgloss.Color("10"))
    StyleSmaller = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
    
    if NoColor {
        lipgloss.ColorOff()
    }

    Result = StyleBigger.Render(myString)

**Additional context**
Of course in my example I only used 2 Styles, imagine it being many more.
And for every new Style I shall not forget to add also the overwriting.

thanks for your feedback upfront.
looking forward to learn how you guys do it
have a great day
--marco

**Comment by testinfected** [comment]
IIUC you can use the following code today to do that:

```go
func DisableColors() {
	os.Setenv("NO_COLOR", "true")
}

func ForceColors() {
	os.Unsetenv("NO_COLOR")
	os.Setenv("CLICOLOR_FORCE", "true")
}

func EnableColors() {
	os.Unsetenv("NO_COLOR")
	os.Unsetenv("CLICOLOR_FORCE")
	os.Setenv("CLICOLOR", "true")
}
```

**Comment by marcotrosi** [comment]
thank you very much @testinfected 
would you be so kind and show me where this is documented?
then I'll be able to find it myself in the future.

**Comment by testinfected** [comment]
Hey @marcotrosi my pleasure

You can find more info here: [Standard for ANSI Colors in Terminals](https://bixense.com/clicolors/)

I had the same issue as you a few months ago and dived into lipgloss source code to see this was properly supported.

**Comment by marcotrosi** [comment]
thanks again @testinfected 
one more question
let's say we have 3 tools in a pipe
`t1 | t2 | t3`
what does it mean for the tools and their output?
especially t3. is t3 in its own sub shell and unaffected or would it have the colors turned off too, as we set the env var?

**Comment by testinfected** [comment]
> thanks again @testinfected one more question let's say we have 3 tools in a pipe `t1 | t2 | t3` what does it mean for the tools and their output? especially t3. is t3 in its own sub shell and unaffected or would it have the colors turned off too, as we set the env var?

I believe the env vars would be inherited by the subshell right? In which case let's that you want to re-enable colors for t3 you would have to use something like:

`t1 | t2 | (unset NOCOLOR; unset CLICOLOR_FORCE; CLICOLOR="true" t3)`

(hope I got the syntax right)

**Comment by testinfected** [comment]
> I believe the env vars would be inherited by the subshell right? In which case let's that you want to re-enable colors for t3 you would have to use something like:
> 
> `t1 | t2 | (unset NOCOLOR; unset CLICOLOR_FORCE; CLICOLOR="true" t3)`

On second thought might not work like this because of the pipe, but I hope you get the idea lol

**Comment by marcotrosi** [comment]
as far as I know on unixoid systems env vars are only inherited to sub shells if one uses the export command

supporting the NO_COLOR solution is definitely a very good and needed feature, but looking at your proposed of changing the environment variables before piping into t3, I must say I would definitely prefer a proper command line parameter for t2

`t1 | t2 --no-color | t3`

basically back to my original request/question.

**Comment by testinfected** [comment]
>  I must say I would definitely prefer a proper command line parameter for t2
> 
> `t1 | t2 --no-color | t3`

Lipgloss doesn't have a CLI tool right?

**Comment by marcotrosi** [comment]
the link you sent @testinfected perfectly states what I meant.
it says ..

_Also consider letting the user override these with command-line options like ... _

**Comment by marcotrosi** [comment]
> > I must say I would definitely prefer a proper command line parameter for t2
> > `t1 | t2 --no-color | t3`
> 
> Lipgloss doesn't have a CLI tool right?

correct, lipgloss is only a lib, but I'm the one who has to take care of the cli options.

**Comment by testinfected** [comment]
Agreed, so we're in the same situation, I add the following to my CLI app:

```bash
--color=auto|always|never
```

Then I use the functions above depending on the flag value.

**Comment by marcotrosi** [comment]
exactly, and now the question if you would have manipulated the behavior of t3 in the example?
See what I mean? It would be horrible if using an option in t2 would affect also t3 in the pipe example. That wouldn't make sense, right?

**Comment by marcotrosi** [comment]
Your link shows it perfectly. Again thanks for sharing it. We need both, NO_COLOR and cli options, but we shouldn't solve the cli options by setting the env vars, this will cause problems from what I see. Do you agree?

**Comment by testinfected** [comment]
Sure does. It's only a workaround for now. I hope #293 will help here

**Comment by meowgorithm** [comment]
Hey everyone. We recommend you have a look at [`charmbracelet/colorprofile`](https://github.com/charmbracelet/colorprofile) to downsample Lip Gloss output (or any ANSI, really) as needed.

It will respect `NO_COLOR`, `CLICOLOR_FORCE` and so on by default, as well as various other environmental factors and also detect if output's not a TTY. It's built on tried and true color downsampling algorithms we've been using for years.

```go
str := someStyle.Render(someContent)

// Detect a color profile and print to stdout, downsampling
// color according to the environment.
w := colorprofile.Detect(os.Stdout, os.Environ())
fmt.Fprintln(w, str)

// Downsample to 4-bit ANSI.
w.Profile = colorprofile.ANSI
fmt.Fprintf(w, str)

// Ascii-fy, no colors.
w.Profile = colorprofile.Ascii
fmt.Fprintf(w, str)

// Strip ANSI altogether.
w.Profile = colorprofile.NoTTY
fmt.Fprintf(w, str)

// Render to a string
str := fmt.Sprint(w, str)
```

In v2 (per #293) there will be wrappers in Lip Gloss for this as well.
